proposals/VK_EXT_map_memory_placed.asciidoc

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 The Khronos Group, Inc.
+//
+// SPDX-License-Identifier: CC-BY-4.0
+
+# VK_EXT_map_memory_placed
+:toc: left
+:refpage: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/
+:sectnums:
+
+This document proposes adding support client-controlled virtual address
+placement of `VkDeviceMemory` maps.
+
+## Problem Statement
+
+There are certain cases in which it can be useful to directly control the
+address used by a memory map operation.  One example of this is in
+emulation environments when the application under emulation is 32-bit but
+the host userspace is 64-bit.  In order to avoid additional copies or
+address translation, the emulator can choose an address in the lower 32
+bits of the address space and request that the map be placed there.
+
+This functionality is already supported by the standard UNIX `mmap()`
+system call which provides an address hint flag as its first parameter as
+well as a `MAP_FIXED` flag which causes the `mmap()` to fail if it cannot
+place the map in exactly that location.  This extension proposes to add
+similar functionality to `vkMapMemory()`.
+
+
+## Solution Space
+
+It is possible to emulate this functionality using
+VK_EXT_external_memory_host by creating a placed memory mapping in some
+other way (such as a Linux memfd combined with conventional `mmap()`) and
+then importing the resulting map via VK_EXT_external_memory_host.  However,
+this requires every mappable Vulkan memory allocation to be a host pointer
+import which places restrictions available memory types and heaps and may
+have significant performance costs.
+
+One alternative approach to the emulation use-case would be to add a
+`VK_MEMORY_MAP_32BIT_ADDRESS_BIT_EXT` which does exactly what emulation
+wants and places the map in the lower 32 bits.  This is equivalent to the
+Linux `MAP_32BIT` flag.  However, this is less flexible and may be
+difficult to implement on Arm platforms where Linux `mmap()` does not
+support `MAP_32BIT` since drivers are likely relying on `mmap()` to
+implement `vkMapMemory()`.
+
+## Proposal
+
+### Dependencies
+
+ - VK_EXT_map_memory2
+
+### API Features
+
+The following features are exposed by this extension:
+
+[source,c]
+----
+typedef enum VkMemoryMapFlagBits {
+    VK_MEMORY_MAP_PLACED_BIT_EXT = 0x00000001,
+} VkMemoryMapFlagBits;
+
+typedef enum VkMemoryUnmapFlagBits {
+    VK_MEMORY_UNMAP_RESERVE_BIT_EXT = 0x00000001,
+} VkMemoryMapFlagBits;
+----
```

Comment:
Is this supposed to be the feature struct here? The placement of these bit flags doesn't make sense to me. You should also illustrate the VkMemoryMapPlacedInfoEXT struct here somewhere in the proposal section.

Reply:
Added all the structs.